grader_systems/src/stage/valgrind.rs

```diff
@@ -0,0 +1,1123 @@
+    config: ValgrindConfig,
+    executable: String,
+    args: Vec<String>,
+    stdin: Option<String>,
+    timeout: Option<Duration>,
+}
+
+fn read_file(path: &Path) -> Result<String> {
+    let mut file = File::open(path)?;
+    let mut contents = String::new();
+    file.read_to_string(&mut contents)?;
+    Ok(contents)
+}
+
+impl<E: ProcessExecutor> Valgrind<E> {
+    #[allow(dead_code)]
+    pub fn new(
+        executor: E,
+        config: ValgrindConfig,
+        executable: String,
+        args: Vec<String>,
+        stdin: Option<String>,
+        timeout: Option<Duration>,
+    ) -> Self {
+        Self {
+            executor,
+            config,
+            executable,
+            args,
+            stdin,
+            timeout,
+        }
+    }
+
+    fn gen_cmd(&self, ws: &Path) -> Command {
+        let mut cmd = Command::new("valgrind");
+
+        cmd.add_arg(format!("--log-file={}", &self.config.log_file));
+
+        if let Some(v) = self.config.leak_check {
+            cmd.add_arg(format!("--leak-check={}", if v { "yes" } else { "no" }));
+        }
+
+        cmd.add_arg(format!("--error-exitcode={}", ERROR_EXITCODE));
+
+        if let Some(v) = self.config.malloc_fill {
+            cmd.add_arg(format!("--malloc-fill=0x{:02X}", v));
+        }
+
+        if let Some(v) = self.config.free_fill {
+            cmd.add_arg(format!("--free-fill=0x{:02X}", v));
+        }
+
+        if let Some(v) = &self.config.suppressions {
+            for supp in v {
+                cmd.add_arg(format!("--suppressions={}", supp));
+            }
+        }
+
+        cmd.add_arg("--");
+        cmd.add_arg(&self.executable);
+        cmd.add_args(&self.args);
+        if let Some(v) = &self.stdin {
+            cmd.set_stdin(StdinPipe::Path(v.into()))
+        }
+
+        cmd.set_timeout(self.timeout.unwrap_or(DEFAULT_TIMEOUT));
+        cmd.set_cwd(ws);
+
+        cmd
+    }
+
+    fn read_logfile(&self, ws: &Path) -> Result<String> {
+        let path = ws.join(&self.config.log_file);
+        if !path.exists() {
+            return Err(anyhow!(
+                "Could not find logfile at {:?}. Did Valgrind actually execute?",
+                path.to_str()
+            ));
+        }
+        let contents = read_file(&path)?;
+        if contents.trim().is_empty() {
+            return Err(anyhow!(
+                "Found empty valgrind log at {:?}. Something went wrong.",
+                path.to_str()
+            ));
+        }
+
+        // replace all absolute paths with basename
+        let re = Regex::new(r"(\W|^)(?:\/[^\/\s]+)+\/([^\/\s]+)\b")?;
+        let repl = re.replace_all(&contents, "$1$2");
+
+        Ok(repl.to_string())
+    }
+}
+
+#[async_trait]
+impl<E: ProcessExecutor> Executor for Valgrind<E> {
+    type Output = StageResult;
+
+    async fn run(&self, ws: &Path) -> Result<Self::Output> {
+        let mut sect = Section::new("Valgrind");
+        let mut results_sect = Section::new("Output:");
```

Comment:
np, remove the ':'. Output will get formatted as a header here which will be bold/bigger than the following text. No need to have the colon.

Reply:
Will do